### PR TITLE
[FEATURE] Port 708 decoder encoding module to RUST

### PR DESCRIPTION
**Pull request by IshanGrover2004**
**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [contributors guide](https://github.com/CCExtractor/ccextractor/blob/master/.github/CONTRIBUTING.md).
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I have considered, and confirmed that this submission will be valuable to others.
- [x] I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
- [x] I give this submission freely, and claim no ownership to its content.
- [x] **I have mentioned this change in the [changelog](https://github.com/CCExtractor/ccextractor/blob/master/docs/CHANGES.TXT).**

**My familiarity with the project is as follows (check one):**

- [ ] I have never used CCExtractor.
- [ ] I have used CCExtractor just a couple of times.
- [ ] I absolutely love CCExtractor, but have not contributed previously.
- [x] I am an active contributor to CCExtractor.

---

I have migrated the logic and functions from the [lib_ccx/ccx_decoder_708_encoding.c](https://github.com/CCExtractor/ccextractor/blob/master/src/lib_ccx/ccx_decoders_708_encoding.c) file to Rust, located in decoder/encoding.rs. This conversion maintains the same functionality that was originally written in C, now implemented in Rust.

To know more See [ccx_decoder_708_encoding.c](https://github.com/CCExtractor/ccextractor/blob/master/src/lib_ccx/ccx_decoders_708_encoding.c) and compare with File Changed in PR, they are both same

### Background & Testing
- **Current codebase scenerio:** Only C code of [ccx_decoder_708_encoding.c](https://github.com/CCExtractor/ccextractor/blob/master/src/lib_ccx/ccx_decoders_708_encoding.c) file is available. So while making & running **Rust** test cases which requires these functions do not have body because `cargo test` is just seeing the rust code only.
- **After PR changes:**  Exact same logic of rust code will be there. And I did also added the conditional compilation so that if user build with `-without-rust` then only C code will work or else Rust code will build. And Yes I have manually tested that when compiling with rust src files, rust code is being used and giving the same result.

### CODE REVIEW

**Review by thealphadollar — APPROVED** [review]
LGTM.

File: src/rust/src/decoder/encoding.rs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+pub extern "C" fn dtvcc_get_internal_from_G2(g2_char: u8) -> u8 {
+    if (0x20..=0x3F).contains(&g2_char) {
+        g2_char - 0x20
+    } else if (0x60..=0x7F).contains(&g2_char) {
+        g2_char + 0x20
+    } else {
+        // Rest unmapped, so we return a blank space
+        0x20
+    }
+}
```
Comment (PunitLodha) [inline src/rust/src/decoder/encoding.rs:36]: Small nitpick but you could use a match here instead

**Review by PunitLodha — COMMENTED** [review]
Are the rust extern functions automatically picked up by C because they have the same name? I remember having to define `extern fn ...` in C and then having to call it like this,
```c
#ifndef DISABLE_RUST
	// Use rust function here
#else
	// Use C function here
#endif
```

Is that not needed?

File: src/rust/src/decoder/encoding.rs (hunk 1 above)
Comment (IshanGrover2004) [reply src/rust/src/decoder/encoding.rs]: Done @PunitLodha

**Comment by PunitLodha** [comment]
@prateekmedia Any idea why Windows is failing here?

**Comment by IshanGrover2004** [comment]
@PunitLodha You can go on for merge if you find correct, btw this method of having same name is working, I have checked it

And another thing I have also Enable the `default_feature` of bindgen bcs this becomes useful while developing in Rust

**Review by PunitLodha — APPROVED** [review]

**Pull request merged**